modin/core/io/file_dispatcher.py

```diff
@@ -152,27 +155,15 @@ def read(cls, *args, **kwargs):
         postprocessing work on the resulting query_compiler object.
         """
         query_compiler = cls._read(*args, **kwargs)
-        # TODO (devin-petersohn): Make this section more general for non-pandas kernel
-        # implementations.
-        if StorageFormat.get() == "Pandas":
-            import pandas as kernel_lib
-        elif StorageFormat.get() == "Cudf":
-            import cudf as kernel_lib
-        else:
-            raise NotImplementedError("FIXME")
-
-        if hasattr(query_compiler, "dtypes") and any(
-            isinstance(t, kernel_lib.CategoricalDtype) for t in query_compiler.dtypes
-        ):
-            dtypes = query_compiler.dtypes
-            return query_compiler.astype(
-                {
-                    t: dtypes[t]
-                    for t in dtypes.index
-                    if isinstance(dtypes[t], kernel_lib.CategoricalDtype)
-                },
-                kwargs.get("errors", "raise"),
-            )
+        # TextFileReader can also be returned from `_read`.
+        if not ExperimentalAsyncReadMode.get() and hasattr(query_compiler, "dtypes"):
+            # at the moment it is not possible to use `wait_partitions` function;
+            # in a situation where the reading function is called in a row with the
+            # same parameters, `wait_partitions` considers that we have waited for
+            # the end of remote calculations, however, when trying to materialize the
+            # received data, it is clear that the calculations have not yet ended.
+            # for example, `test_io_exp.py::test_read_evaluated_dict` is failed because of that
+            _ = query_compiler.dtypes
```

Comment:
Not sure I fully understand this.

1. Why do we need to call dtypes?
2. When you are saying about using `wait_partitions` and materializing the received data, why is it clear that the calculations have not yet ended?

Reply:
> 1. Why do we need to call dtypes?

This is a safer way to wait for calculations in a remote function.

> 2. When you are saying about using wait_partitions and materializing the received data, why is it clear that the calculations have not yet ended?

I observed the following situation:

```python
import time
# add time.sleep(20) for the parser, AsyncReadMode is on
start = time.time()
pd.read_custom_text(args)
print(f"first call time: {time.time()-start}") > 20 sec

start = time.time()
pd.read_custom_text(args) # same args
print(f"second call time: {time.time()-start}") < 1 sec
```

Reply:
Interesting, can you please raise an issue regarding this so we can investigate this behavior further? Maybe the first call somehow caches the result and then returns it when we being called for the second time...

Reply:
https://github.com/modin-project/modin/issues/5944